### ISSUE TITLE
[plugin] deal correctly with recursive symlinks and add a unit test for them

### ISSUE THREAD

**Issue by bmr-cymru** [sosreport/sos#567]
It was pointed out on the sos-devel list that recursive symlinks (i.e. a symlink containing its own path as target) are not properly handled by sos. This leads to an exception during archive IO and abortion of collection:

```
# sosreport -vv --batch --debug -o general

sosreport (version 3.2)

[...]

[plugin:general] recursively adding 'foo' from '/etc/sysconfig'
[plugin:general] copying link '/etc/sysconfig/foo' pointing to 'foo' with isdir=False
[archive:TarFileArchive] added symlink at '/var/tmp/sosreport-bmr-rhel7-vm1-20150526154403/etc/sysconfig/foo' to 'foo' in FileCacheArchive '/var/tmp/sosreport-bmr-rhel7-vm1-20150526154403'
[plugin:general] normalized link target 'foo' as '/etc/sysconfig/foo'
[plugin:general] copying link '/etc/sysconfig/foo' pointing to 'foo' with isdir=False

{{  perfectly silent BOOM! as the exception is caught by the generic FS error handler }}
```

Obviously this needs fixing but we should also have a case for this in the unit suite.

**Comment by pierg75** [comment]
Is this still open?
In a quick test I see that this should be taken care by the commit ab55d6051:

```
ab55d605 [plugin] fix recursive symlink handling
diff --git a/sos/plugins/__init__.py b/sos/plugins/__init__.py
index e50e0d31..21e01867 100644
--- a/sos/plugins/__init__.py
+++ b/sos/plugins/__init__.py
@@ -268,7 +268,13 @@ class Plugin(object):
         # to absolute paths to pass to _do_copy_path.
         self._log_debug("normalized link target '%s' as '%s'"
                         % (linkdest, absdest))
-        self._do_copy_path(absdest)
+
+        # skip recursive copying of symlink pointing to itself.
+        if (absdest != srcpath):
+            self._do_copy_path(absdest)
+        else:
+            self._log_debug("link '%s' points to itself, skipping target..."
+                            % linkdest)
 
         self.copied_files.append({'srcpath': srcpath,
                                   'dstpath': srcpath,

```

Does this only need a unittest (I couldn't see one for this use case)?

**Comment by bmr-cymru** [comment]
Yeah: just the unit test now - it's probably not that complicated. The main thing will be creating the self-referencing link and then cleaning it up properly (it doesn't seem a good idea to leave those in the source tree).

**Comment by pierg75** [comment]
Would something like in the pull request #1135 work?

**Comment by TurboTurtle** [comment]
I believe this is resolved via 8e60e29, so closing this out.

If this is not the case, please feel free to re-open this issue.
